media/windows/ECC-Ecobee/ECC Pycobee Data.py

```diff
@@ -1019,22 +1088,38 @@ def main():
         F"Total snapshot database records written, all other tables (than runtime):  {snapshot_recs_written_total}")
     print(F"Total snapshot database records written, all other tables (than runtime):  {snapshot_recs_written_total}")
 
+    # Calculate and display the database size increase this execution, in bytes...
+    db_bytes_increase = db_close_size_bytes - db_open_size_bytes
+    logger.info(F"Database size increased by {db_bytes_increase} bytes...")
+    print(F"\nDatabase size increased by {db_bytes_increase} bytes...")
     now = datetime.now()
     date_now_str = now.strftime("%Y-%m-%d %H:%M:%S")
     print(F"*** Execution completed at:  {date_now_str} ***")
     logger.info(F"*** Execution completed at:  {date_now_str} ***")
 
 
-def persist_to_shelf(file_name, ecobee_service):
-    pyecobee_db = shelve.open(file_name, protocol=4)
-    pyecobee_db['application_key'] = ecobee_service.application_key
-    logger.debug(F"Persist access token:  {ecobee_service.access_token}")
-    logger.debug(F"Persist refresh token:  {ecobee_service.refresh_token}")
-    pyecobee_db['access_token'] = ecobee_service.access_token
-    pyecobee_db['refresh_token'] = ecobee_service.refresh_token
-    pyecobee_db['authorization_token'] = ecobee_service.authorization_token
-
-    pyecobee_db.close()
+def persist_to_json(auth_json_file_name, ecobee_service):
+    # json_auth_dict = {}
+    try:
+        with open(auth_json_file_name, "w") as write_auth:
+            json_auth_dict['application_key'] = ecobee_service.application_key
+            logger.debug(F"Persist access token:  {ecobee_service.access_token}")
+            logger.debug(F"Persist refresh token:  {ecobee_service.refresh_token}")
+            # print(F"Persist access token:  {ecobee_service.access_token}")
+            # print(F"Persist refresh token:  {ecobee_service.refresh_token}")
+            json_auth_dict['access_token'] = ecobee_service.access_token
+            json_auth_dict['refresh_token'] = ecobee_service.refresh_token
+            json_auth_dict['authorization_token'] = ecobee_service.authorization_token
+    
+            json.dump(json_auth_dict, write_auth)
+    except Exception as e:
+        logger.error(F"Error occurred while attempting to write JSON tokens file...{e}")
+        logger.error(F"...aborting...")
+        print(F"Error occurred while attempting to write JSON tokens file...{e}")
+        print(F"...aborting...")
+        sys.exit(1)
+ 
+    write_auth.close()
```

Comment:
`write_auth` will already be closed by this point.  With this syntax in Python:

```python
with open(...) as blah:
    ...do something with blah...
# when python reaches here (i.e., outside the "with" scope), 
# the file has guaranteed to be closed.
```

Reply:
Yeah, good catch.  This was the first time I'd used it and wasn't sure if that was the case so I made it explicit.  I think I have a more recent use where I omitted it.  I'll drop that line.

Reply:
Yeah, the `with` statement is super cool.  Most intrinsic Python entities have Context Managers associated with them (including file handles), so that when used in a `with` block, it guarantees to do cleanup upon exit of the block, even if an exception occurs inside the block.  Check out:

* https://docs.python.org/3/reference/compound_stmts.html#with and
* https://docs.python.org/3/reference/datamodel.html#context-managers